Add AWS CDK Infrastructure for Seatool Connectors

benjaminpaige opened the pull request:
# 🏗️ Add initial AWS CDK Infrastructure for Seatool Connectors

## Overview

This PR introduces AWS CDK (Cloud Development Kit) infrastructure to the Seatool Connectors project, establishing the foundation for migrating from Serverless Framework to native AWS CloudFormation through CDK. This change provides better infrastructure as code practices, enhanced type safety, and improved resource management capabilities.

## 📊 Changes Summary

- **18 files changed**: 3,457 insertions, 27 deletions
- **New CDK infrastructure**: Complete alerts stack with SNS topics and KMS encryption
- **GitHub Actions workflows**: Automated deployment and destruction pipelines
- **Configuration system**: Environment-based configuration with SSM parameter management
- **Enhanced testing**: Improved coverage reporting and CDK-aware test configuration

## 🎯 Key Features

### 1. CDK Infrastructure Foundation
- **Alerts Stack**: SNS topics with KMS encryption for ECS task failure notifications
- **Shared utilities**: VPC lookup, SSM parameter management, and environment configuration
- **Custom bootstrap qualifier**: `seatool` namespace to prevent resource conflicts with other CDK projects

### 2. Environment Configuration System
- **Multi-environment support**: `dev`, `val`, `master`, and `production` stages
- **Flexible configuration**: CPU, memory, IAM, VPC, and alerting settings per environment
- **SSM parameter integration**: Cross-stack communication and configuration sharing

### 3. Automated Deployment Pipeline
- **Deploy workflow**: Automatic deployment on CDK file changes with security scanning
- **Destroy workflow**: Manual infrastructure teardown with environment cleanup
- **Bootstrap automation**: Intelligent CDK bootstrap with custom qualifier
- **Concurrency control**: Prevents conflicting deployments per stage

### 4. Security & Compliance
- **KMS encryption**: All SNS topics encrypted with dedicated KMS keys
- **IAM policies**: Least-privilege access with proper service principal permissions
- **Termination protection**: Enabled for production environments
- **Security scanning**: cfn_nag integration for CloudFormation template validation

## 🚀 Migration Strategy

This CDK implementation begins with the alerts infrastructure and sets up the foundation for migrating additional stacks:

**Current**: 
- ✅ Alerts Stack (SNS topics, KMS keys, CloudWatch integration)

**Planned**:
- Dashboard Stack
- Connector Stack  
- Debezium Stack
- KsqlDB Stack
- KsqlThree Stack

## 🔧 Technical Implementation

### Stack Architecture
- **Custom construct IDs**: Consistent naming with environment prefixes
- **Resource tagging**: Automatic PROJECT and SERVICE tags for cost tracking
- **Export parameters**: SSM parameters for cross-stack resource sharing
- **CloudFormation outputs**: Stack exports for external integrations

### Configuration Highlights
```typescript
// Environment-specific configuration
export interface EnvironmentConfig {
  account?: string;
  region: string;
  project: string;
  stage: string;
  topicNamespace: string;
  terminationProtectionEnabled: boolean;
  // ... additional environment settings
}
```

### Resource Naming Convention
- **Stacks**: `{project}-cdk-{service}-{stage}` (e.g., `seatool-cdk-alerts-dev`)
- **Resources**: `{project}-{service}-{resource}-{stage}` (e.g., `seatool-alerts-ecs-failure-dev`)
- **Constructs**: `{service}{Resource}{stage}` (e.g., `alertsEcsFailureDev`)

## 📋 Deployment Instructions

### Prerequisites
- Node.js v20.17.0
- AWS CDK CLI v2.179.0
- AWS credentials configured

### Local Development
```bash
# Setup and build
yarn install
yarn cdk:build

# Deploy infrastructure
export STAGE=dev
yarn cdk:deploy

# Destroy infrastructure
yarn cdk:destroy
```

### GitHub Actions
- **Automatic**: Push to any branch with CDK changes
- **Manual**: Workflow dispatch with custom stage parameter
- **Security**: cfn_nag scanning and notifications on failure

## 🧪 Testing Enhancements

- **Coverage reporting**: Enhanced Vitest configuration with detailed coverage output
- **Code Climate integration**: Automatic coverage reporting to external services
- **CDK exclusions**: TypeScript configuration updated to exclude CDK compilation

## 📚 Documentation

Comprehensive documentation added in `cdk/README.md` covering:
- Architecture overview and migration roadmap
- Local development setup and commands
- GitHub Actions workflow details
- Environment configuration guide
- Troubleshooting and common issues

## 🔄 Breaking Changes

**None** - This is purely additive infrastructure. Existing Serverless Framework stacks remain unchanged and operational.

## 🔍 Files Changed

### New CDK Infrastructure
- `cdk/` - Complete CDK application with TypeScript configuration
- `cdk/lib/stacks/alerts-stack.ts` - SNS topics and KMS infrastructure
- `cdk/lib/shared/` - Shared utilities for configuration and SSM parameters

### GitHub Actions
- `.github/workflows/deploy-cdk.yml` - CDK deployment automation
- Updates to destroy and unit test workflows

### Configuration Updates
- `package.json` - CDK dependencies and scripts
- `tsconfig.json` - CDK file exclusions
- `.gitignore` - CDK output directories

## 🎉 Benefits

1. **Type Safety**: TypeScript-based infrastructure with compile-time validation
2. **Resource Management**: Better resource organization and naming consistency
3. **Security**: KMS encryption and IAM best practices built-in
4. **Automation**: Streamlined deployment with GitHub Actions integration
5. **Scalability**: Foundation for migrating remaining Serverless Framework stacks
6. **Observability**: Enhanced monitoring and alerting capabilities

## 🔗 Related Issues

This PR establishes the infrastructure foundation for future serverless service migrations and improved operational capabilities.

---

**Reviewer Notes**: This PR introduces significant infrastructure changes. Please review the CDK configuration, GitHub Actions workflows, and documentation for completeness and security considerations.

github-actions[bot] committed:
Release to val (#163)

* Create dsacms.yml

* Update dsacms.yml

* Update code.json (#164)

* Update code.json metadata

* update update-code-json

* update to code.json

---------

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>
Co-authored-by: Benjamin Paige <benjamin.paige@gmail.com>

* fix: downsize ksqlthree cpu and memory usage per profesormac (#166)

* fix: downsize ksqlthree cpu and memory usage per profesormac

* fix: cpu and memory optimization and cost reduction

---------

Co-authored-by: Seun0925 <ssoyinka@MacBook-Pro-2.local>

---------

Co-authored-by: Benjamin Paige <benjamin.paige@gmail.com>
Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>
Co-authored-by: Seun0925 <166032098+Seun0925@users.noreply.github.com>
Co-authored-by: Seun0925 <ssoyinka@MacBook-Pro-2.local>

benjaminpaige committed:
feat: add AWS CDK infrastructure for Seatool Connectors

- Updated .gitignore to exclude CDK generated files.
- Added new scripts in package.json for CDK installation, build, synthesis, deployment, and destruction.
- Created GitHub Actions workflow for deploying CDK infrastructure.
- Introduced destroy workflow for managing CDK resources.
- Added CDK configuration files and shared utilities for environment management.
- Implemented Alerts stack with SNS and KMS integration.
- Established shared configuration and SSM parameter management for cross-stack communication.

benjaminpaige committed:
chore: update tsconfig.json to exclude CDK files

- Added "./cdk/**/*" to the exclude list in tsconfig.json to prevent TypeScript from compiling CDK-related files.

macpro-snyk-service-account commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **code/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/macpro-macpro/project/21e9cf02-f6ae-4f5d-8458-1b2de4652c4c/pr-check/24a6c9a8-4ace-454c-b2c1-ceedada63e4a?refs=6bf539f765a164e82423f07732369e84f4f88fda&source=prChecksComment)

benjaminpaige committed:
chore: update CDK bootstrap process and documentation

- Modified the CDK bootstrap check to use a custom SSM parameter for determining if bootstrapping is needed, enhancing resource management.
- Updated the CDK configuration to include a custom bootstrap qualifier (`seatool`) for namespacing resources.
- Added documentation on the custom bootstrap qualifier to the README for clarity on resource management in multi-project environments.

benjaminpaige committed:
chore: enhance testing configuration and CDK context

- Updated the coverage script in package.json to include a specific config file for Vitest.
- Enhanced the Vitest configuration to specify coverage options, including report formats and exclusions.
- Moved the bootstrap qualifier for CDK into the context section for better organization and clarity.

benjaminpaige committed:
chore: refine CDK bootstrap process and enhance test coverage reporting

- Updated the CDK bootstrap check to improve clarity by reversing the condition and adding a specific toolkit stack name.
- Enhanced the unit test workflow to include detailed coverage reporting, ensuring the coverage file is checked for existence and content before publishing to Code Climate.

benjaminpaige committed:
chore: refine CDK bootstrap process and enhance test coverage reporting

- Updated the CDK bootstrap check to improve clarity by reversing the condition and adding a specific toolkit stack name.
- Enhanced the unit test workflow to include detailed coverage reporting, ensuring the coverage file is checked for existence and content before publishing to Code Climate.

benjaminpaige committed:
refactor: optimize CPU and memory configurations for ksqldb and ksqlthree services

- Adjusted resource allocations for ksqldb:
  - Reduced heap to 6GB and memory to 12GB for better fit in container.
  - Set RocksDB cache to 1GB.

- Updated ksqlthree configurations:
  - Increased master and production CPU to 4 vCPU and heap to 8GB for improved performance.
  - Set validation environment to 2 vCPU and 5GB heap, optimizing resource usage.

These changes enhance performance and stability while ensuring efficient resource utilization across environments.

benjaminpaige committed:
feat: add comprehensive migration plan from Serverless Framework to AWS CDK

- Introduced a detailed migration strategy for the `seatool-connectors` project, ensuring exact resource parity and zero downtime.
- Documented service dependencies, architecture, and environment stages.
- Outlined a phased migration approach with specific steps for pre-migration setup, service migration order, and validation processes.
- Included risk mitigation strategies, rollback procedures, and post-migration activities to ensure a smooth transition.
- Provided success criteria and monitoring strategies to validate migration effectiveness.

benjaminpaige committed:
chore: update issue templates for bug, enhancement, and question submissions

- Revised the bug template to improve clarity and structure, adding sections for impact, reproduction steps, and acceptance criteria.
- Enhanced the enhancement template with clearer sections for feature description, proposed solutions, and business value.
- Updated the question template to include context, previous attempts, and desired outcomes, ensuring users provide comprehensive information.

These changes aim to streamline issue reporting and improve the quality of information provided by users.

benjaminpaige committed:
chore: update package configurations and dependencies across services

- Added versioning to package.json files for libs and dashboard services.
- Updated runtime from nodejs18.x to nodejs20.x in serverless.yml files for alerts, connector, dashboard, debezium, ksqldb, and ksqlthree services.
- Updated AWS SDK dependencies for ksqldb and ksqlthree services to the latest versions.

These changes ensure compatibility with the latest Node.js version and improve dependency management across the project.

benjaminpaige committed:
chore: update TypeScript and type definitions in package configurations

- Added TypeScript version 5.0.0 to the dashboard service and main package.json.
- Included type definitions for minimatch and node to enhance type safety and compatibility.

These updates improve type checking and ensure alignment with the latest TypeScript features.

dependabot[bot] committed:
chore(deps): bump axios in the npm_and_yarn group across 1 directory (#180)

Bumps the npm_and_yarn group with 1 update in the / directory: [axios](https://github.com/axios/axios).


Updates `axios` from 1.7.9 to 1.12.2
- [Release notes](https://github.com/axios/axios/releases)
- [Changelog](https://github.com/axios/axios/blob/v1.x/CHANGELOG.md)
- [Commits](https://github.com/axios/axios/compare/v1.7.9...v1.12.2)

---
updated-dependencies:
- dependency-name: axios
  dependency-version: 1.12.2
  dependency-type: indirect
  dependency-group: npm_and_yarn
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

benjaminpaige committed:
chore: update Code Climate action to v6.0.0 and enable debug mode

- Upgraded the Code Climate action version from v3.2.0 to v6.0.0 in the unit tests workflow.
- Enabled debug mode for improved troubleshooting during test coverage publishing.

These changes enhance the testing workflow and provide better insights during coverage reporting.

benjaminpaige committed:
chore: temporarily disable Code Climate coverage reporting due to download issues

- Updated the unit tests workflow to disable the Code Climate coverage reporting step.
- This change is a temporary measure to address ongoing download issues with Code Climate.

These adjustments ensure the workflow continues to function while resolving external dependencies.
